### ISSUE TITLE
Bug in hash function used to reduce results data transferred over network

### ISSUE THREAD

**Issue by G-200** [TelegramMessenger/Telegram-iOS#1161]
Here hash function specified: https://core.telegram.org/api/offsets#hash-generation

```
hash = 0
for id in ids:
    hash = hash ^ (id >> 21)
    hash = hash ^ (id << 35)
    hash = hash ^ (id >> 4)
    hash = hash + id
```

But in code its implementation is incorrect:

https://github.com/TelegramMessenger/Telegram-iOS/blob/a196a1684b5c2afeea249aaf8c6b668fdb74927b/submodules/TelegramCore/Sources/State/HistoryViewStateValidation.swift#L421-L426

But the problem seems to be even deeper. I tried to fix it in my fork, but server still returns hashes that are different from those calculated on client with the described hash function.

For example messages.getStickers still returns full result each time, instead of stickersNotModified.

Can you look into this and synchronize hash function used on client and server?

**Comment by ali-fareed** [comment]
Thank you for the observation, the hash function at https://core.telegram.org/api/offsets#hash-generation contains a typo. It should be `hash = hash ^ (hash >> 21)` and so on (obviously, `id >> 21` doesn't make any sense in the context of hashing).

Do you have a specific example where `messages.getStickers` returns a full result when you pass it a `hash` calculated via `combineInt64Hash`?

**Comment by ali-fareed** [comment]
It is a bug in the official Android client (it has been reported internally).

```
hash = hash ^ (hash >> x)
...
hash = hash + next
```

is equivalent to

```
hash = (hash ^ next) + next
```

which is not a great hash function.

On the iOS side, I'm seeing `featuredStickersNotModified` as expected. Could you share your `category` and `hash`?

**Comment by ali-fareed** [comment]
You are right, the iOS app does not account for unseen sticker packs, even though it should. The `hashForIdsReverse ` function should be as follows:

```swift
private func hashForIdsReverse(_ ids: [Int64], unreadIds: [Int64]) -> Int64 {
    var acc: UInt64 = 0
    
    for id in ids {
        combineInt64Hash(&acc, with: UInt64(bitPattern: id))
        if unreadIds.contains(id) {
            combineInt64Hash(&acc, with: 1 as UInt64)
        }
    }
    return Int64(bitPattern: acc)
}
```

**Comment by ali-fareed** [comment]
Fixed in 10.0.0.
